src/util/line-monitor.ts

```diff
@@ -0,0 +1,72 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {Deferred} from './deferred.js';
+
+import type {ScriptChildProcess} from '../script-child-process.js';
+import type {Result} from '../error.js';
+
+/**
+ * Monitors the stdout and stderr of a child process line-by-line searching for
+ * a match of the given regular expression.
+ */
+export class LineMonitor {
+  private readonly _child: ScriptChildProcess;
+  private readonly _pattern: RegExp;
+  private readonly _matched = new Deferred<Result<void, void>>();
+  private _stdout = '';
+  private _stderr = '';
+
+  /**
+   * Resolves to `{"ok": true}` when a match was found or `{"ok": false}` when
+   * this monitor was aborted.
+   */
+  readonly matched = this._matched.promise;
+
+  constructor(child: ScriptChildProcess, pattern: RegExp) {
+    this._child = child;
+    this._pattern = pattern;
+    child.stdout.on('data', this._onStdout);
+    child.stderr.on('data', this._onStderr);
+  }
+
+  abort() {
+    this._removeEventListeners();
+    this._matched.resolve({ok: false, error: undefined});
+  }
+
+  private _removeEventListeners() {
+    this._child.stdout.removeListener('data', this._onStdout);
+    this._child.stderr.removeListener('data', this._onStderr);
+  }
+
+  private _onStdout = (data: string | Buffer) => {
+    this._stdout = this._check(this._stdout + String(data));
+  };
+
+  private _onStderr = (data: string | Buffer) => {
+    this._stderr = this._check(this._stderr + String(data));
+  };
+
+  private _check(buffer: string): string {
```

Comment:
Nit: Could this manual buffering be replaced by using Node's `readline` module, which is supported for Node 14 and up.
E.g.: https://nodejs.org/docs/latest-v14.x/api/readline.html#readline_readline

This then has an event caused `line`: https://nodejs.org/docs/latest-v14.x/api/readline.html#readline_event_line which will provide the single line.

Reply:
I don't think so, because readline only gives you a line when it completed (so only when `\n` is printed do you get the previous line). But we want to check even an incomplete line, whenever any output is emitted.

Because imagine a script that just writes "Listening on port 8080" without writing a trailing newline.

Updated the test to cover that case, and added a comment.

Reply:
Brilliant!